compiler/rustc_mir_transform/src/dest_prop/mod.rs

```diff
@@ -0,0 +1,1011 @@
+//!
+//! # Motivation
+//!
+//! MIR building can insert a lot of redundant copies, and Rust code in general often tends to move
+//! values around a lot. The result is a lot of assignments of the form `dest = {move} src;` in MIR.
+//! MIR building for constants in particular tends to create additional locals that are only used
+//! inside a single block to shuffle a value around unnecessarily.
+//!
+//! LLVM by itself is not good enough at eliminating these redundant copies (eg. see
+//! <https://github.com/rust-lang/rust/issues/32966>), so this leaves some performance on the table
+//! that we can regain by implementing an optimization for removing these assign statements in rustc
+//! itself. When this optimization runs fast enough, it can also speed up the constant evaluation
+//! and code generation phases of rustc due to the reduced number of statements and locals.
+//!
+//! # The Optimization
+//!
+//! Conceptually, this optimization is "destination propagation". It is similar to the Named Return
+//! Value Optimization, or NRVO, known from the C++ world, except that it isn't limited to return
+//! values or the return place `_0`. On a very high level, independent of the actual implementation
+//! details, it does the following:
+//!
+//! 1) Identify `dest = src;` statements with values for `dest` and `src` whose storage can soundly
+//!    be merged.
+//! 2) Replace all mentions of `src` with `dest` ("unifying" them and propagating the destination
+//!    backwards).
+//! 3) Delete the `dest = src;` statement (by making it a `nop`).
+//!
+//! Step 1) is by far the hardest, so it is explained in more detail below.
+//!
+//! ## Soundness
+//!
+//! Given a pair of places `dest` and `src` whose memory we would like to merge, we need to check
+//! the following conditions:
+//!
+//! * `dest` and `src` must both be *constant*. This is satisfied if they do not contain any
+//!   indirection through a pointer or any indexing projections.
+//!
+//! * If `dest` and `src` are to use the same memory, it must be possible for them to "split up
+//!   ownership" of that memory in a clean way. More precisely, we need to be able to partition the
+//!   CFG into two disjoint parts, so that `dest` can have the memory during one part and `src` can
+//!   have the memory during the other part. We will choose these parts to be the [live
+//!   ranges][liveness] of the place. This gets us most of the way there, since if the live ranges
+//!   do not overlap it means that the places are never trying to simultaneously use the memory to
+//!   store a value that they'll need again later. However, we additionally need to ensure that
+//!   neither place interferes with the other's usage of the memory. This means we must eliminate
+//!   all stores to the memory that take place outside of the place's live range - this is typically
+//!   a separate optimization known as "dead store elimination".
```

Comment:
If a local variable is modified in a statement or terminator its use will be in conflict with any variables that are live at the time (with the exception of assignments that form the basis for the propagation).

Would it be possible to check this condition directly and remove dependency on dead store elimination? 

The coupling with the dead store elimination feels very subtle and will likely go out of sync (in fact it already did with a variant of it that lands separately, since it no longer removes all dead assignments).

Reply:
Yes, this was my plan too, for all the reasons you stated. Hoping to get a chance to rebase this in the next few days

Reply:
I've implemented this locally and completed a full rebase. I've also restructured much of the code to make it much more clear. Tomorrow is documentation and test day